### PR TITLE
Prevent Forms from introducing dataset name case conflicts

### PR DESCRIPTION
**Pull request by ktuite**
Closes https://github.com/getodk/central/issues/627

Prevents Forms and Dataset API from introducing published datasets with conflicting names, specifically the same name with different capitalization. 

* Error on _Form Upload_ if dataset name conflicts with a published dataset
* Error on _Form Publish_ if publishing will introduce a conflict, which can happen if there were multiple draft forms with different name variations, one was published, and then a second is attempted to be published.
* Error on "New Entity List" / API if dataset name conflicts with a published dataset
   * Now allowed to create/publish a dataset that directly collides with an unpublished dataset

#### What has been done to verify that this works as intended?

Tests, trying with frontend.

#### Why is this the best possible solution? Were any other approaches considered?

I basically added another database query for a similarly named dataset. In the same area of code, we're also querying for an existing exactly matching dataset so we can reuse its ID and Actee ID for various things. I wasn't comfortable trying to combine these queries at this time, given the complicated logic of draft datasets, similar but not exact matches, multiple similar draft datasets, etc. 

#### Does this change require updates to the API documentation? If so, please update docs/api.yaml as part of this PR.

This could potentially be added to the API docs. (https://github.com/getodk/central-backend/issues/1096)

#### Before submitting this PR, please make sure you have:

- [ ] run `make test` and confirmed all checks still pass OR confirm CircleCI build passes
- [ ] verified that any code from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

File: lib/model/query/datasets.js [hunk 1]
```diff
@@ -153,41 +179,63 @@ createPublishedProperty.audit = (property, dataset) => (log) =>
...
+
+  const _datasetNameConflict = () => sql`
+    SELECT existing_ds.name as current, ds_to_publish.name as provided FROM dataset_form_defs dfd
+    JOIN datasets ds_to_publish ON dfd."datasetId" = ds_to_publish.id
+    JOIN datasets existing_ds ON LOWER(ds_to_publish.name) = LOWER(existing_ds.name)
+      AND ds_to_publish.id != existing_ds.id
+      AND existing_ds."publishedAt" IS NOT NULL
+      AND ds_to_publish.name != existing_ds.name
+    WHERE dfd."formDefId" = ${formDefId}
+      AND ds_to_publish."publishedAt" IS NULL
```
Comment (ktuite) [inline lib/model/query/datasets.js:229]: Join on `existing_ds` needs to filter by `projectId`

File: lib/model/query/datasets.js [hunk 2]
```diff
@@ -153,41 +179,63 @@ createPublishedProperty.audit = (property, dataset) => (log) =>
...
+  ORDER BY "propertyId"
+  `;
+
+  const _datasetNameConflict = () => sql`
+    SELECT existing_ds.name as current, ds_to_publish.name as provided FROM dataset_form_defs dfd
+    JOIN datasets ds_to_publish ON dfd."datasetId" = ds_to_publish.id
+    JOIN datasets existing_ds ON LOWER(ds_to_publish.name) = LOWER(existing_ds.name)
+      AND ds_to_publish.id != existing_ds.id
+      AND existing_ds."publishedAt" IS NOT NULL
+      AND ds_to_publish.name != existing_ds.name
```
Comment (ktuite) [inline lib/model/query/datasets.js:227]: Condition not necessary, dataset already exists at this point so it wont be possible for two with the same name to already exist.

File: lib/model/query/datasets.js [hunk 3]
```diff
@@ -153,41 +179,63 @@ createPublishedProperty.audit = (property, dataset) => (log) =>
...
+    WHERE dfd."formDefId" = ${formDefId}
+      AND ds_to_publish."publishedAt" IS NULL
+  `;
+
+  const d = await maybeOne(_datasetNameConflict(formDefId));
```
Comment (ktuite) [inline lib/model/query/datasets.js:232]:
```suggestion
  const d = await maybeOne(_datasetNameConflict());
```

File: lib/model/query/datasets.js [hunk 4]
```diff
@@ -153,41 +179,63 @@ createPublishedProperty.audit = (property, dataset) => (log) =>
...
+  if (d.isDefined()) {
+    const { current, provided } = d.get();
+    throw Problem.user.datasetNameConflict({ current, provided });
+  }
+  return all(_publishIfExists(formDefId, publishedAt));
```
Comment (ktuite) [inline lib/model/query/datasets.js:237]:
```suggestion
  return all(_publishIfExists());
```

File: lib/model/query/datasets.js [hunk 5]
```diff
@@ -123,7 +131,25 @@ const createOrMerge = (parsedDataset, form, fields) => async ({ one, Actees, Dat
...
 // Insert a Dataset when there is no associated form and immediately publish
-const createPublishedDataset = (dataset, project) => async ({ one, Actees }) => {
+const createPublishedDataset = (dataset, project) => async ({ one, Actees, Datasets }) => {
+
+  // Check for existing dataset with the same name
+  const existingDataset = await Datasets.get(project.id, dataset.name, false);
+  if (existingDataset.isDefined()) {
+    const { publishedAt, id: datasetId } = existingDataset.get();
+    if (publishedAt)
+      throw Problem.user.uniquenessViolation({ table: 'datasets', fields: [ 'name', 'projectId' ], values: [ dataset.name, project.id ], });
```
Comment (ktuite) [inline lib/model/query/datasets.js:141]:
```suggestion
      throw Problem.user.uniquenessViolation({ table: 'datasets', fields: [ 'name', 'projectId' ], values: [ dataset.name, project.id ] });
```

File: lib/model/query/datasets.js [hunk 6]
```diff
@@ -123,7 +131,25 @@ const createOrMerge = (parsedDataset, form, fields) => async ({ one, Actees, Dat
...
+
+    // If existing dataset is only a draft, allow it to be published
+    return new Dataset(await one(sql`UPDATE datasets SET "publishedAt" = "createdAt" where id = ${datasetId} RETURNING *`));
```
Comment (ktuite) [inline lib/model/query/datasets.js:144]: Set `publishedAt` to current clock timestamp rather than `createdAt`

File: lib/model/query/datasets.js [hunk 7]
```diff
@@ -123,7 +131,25 @@ const createOrMerge = (parsedDataset, form, fields) => async ({ one, Actees, Dat
...
+  }
+
+  // Check for a published dataset with a similar name but different capitalization
+  const conflictingName = await Datasets.getPublishedBySimilarName(project.id, dataset.name);
+  if (conflictingName.isDefined())
+    throw Problem.user.datasetNameConflict({ current: conflictingName.get().name, provided: dataset.name });
```
Comment (ktuite) [inline lib/model/query/datasets.js:150]: Move this above. Situation:

draft "people"
draft "PEOPLE"
publish "people"
API to create "PEOPLE" (should not be allowed)

File: lib/util/problem.js [hunk 8]
```diff
@@ -206,7 +206,10 @@ const problems = {
 
     // 409.14 (entityViolation) was used in the past, but has been deprecated
 
-    entityVersionConflict: problem(409.15, ({ current, provided }) => `Current version of the Entity is '${current}' and you provided '${provided}'. Please correct the version number or pass '?force=true' in the URL to forcefully update the Entity.`)
+    entityVersionConflict: problem(409.15, ({ current, provided }) => `Current version of the Entity is '${current}' and you provided '${provided}'. Please correct the version number or pass '?force=true' in the URL to forcefully update the Entity.`),
+
+    datasetNameConflict: problem(409.16, ({ current, provided }) => `A dataset named '${current}' exists and you provided '${provided}' with the same name but different capitalization. Please change the capitalization of '${provided}' to match '${current}'.`)
```
Comment (ktuite) [inline lib/util/problem.js:211]: Remove second part of sentence to make it make more sense for the new Entity List via API modal.

File: test/integration/api/datasets.js [hunk 9]
```diff
@@ -98,22 +98,81 @@ describe('datasets and entities', () => {
...
-      }));
+          // Second time
+          await asAlice.post('/v1/projects/1/datasets')
+            .send({
+              name: 'trees'
+            })
+            .expect(409)
+            .then(({ body }) => {
+              body.code.should.equal(409.3);
+              body.message.should.startWith('A resource already exists with name,projectId value(s) of trees,');
```
Comment (ktuite) [inline test/integration/api/datasets.js:119]: Can add projectID 1 back

File: test/integration/api/datasets.js [hunk 10]
```diff
@@ -3464,6 +3523,180 @@ describe('datasets and entities', () => {
...
+          await alice.post('/v1/projects/1/forms/simpleEntity/draft/publish')
+            .expect(200);
+
+          await alice.post('/v1/projects/1/forms/simpleEntity2/draft/publish')
+            .expect(409)
+            .then(({ body }) => {
+              body.code.should.equal(409.16);
+              body.message.should.startWith("A dataset named 'people' exists and you provided 'PEOPLE'");
+            });
+        }));
```
Comment (ktuite) [inline test/integration/api/datasets.js:3611]: Check path of publishing a form while also specifying a new version

File: test/integration/api/datasets.js [hunk 11]
```diff
@@ -3464,6 +3523,180 @@ describe('datasets and entities', () => {
...
+            .expect(200);
+
+          // publish "people"
+          await alice.post('/v1/projects/1/forms/simpleEntity/draft/publish')
+            .expect(200);
+
+          // dataset "people" (same as first/published)
+          await alice.post('/v1/projects/1/forms')
+            .send(testData.forms.simpleEntity
+              .replace('dataset="people"', 'dataset="people"')
```
Comment (ktuite) [inline test/integration/api/datasets.js:3693]: dont need to replace people

**Review by ktuite — COMMENTED** [review]
Interactive review with @matthew-white

**Review by matthew-white — APPROVED** [review]

**Pull request merged**